src/main/java/com/facebook/openwifirrm/modules/ModelerUtils.java

```diff
@@ -381,4 +386,157 @@ public static Map<String, Map<String, WifiScanEntry>> getAggregatedWifiScans(
+					String key = String.format(
+						"bssid: %s, station: %s",
+						association.bssid,
+						association.station
+					);
+					List<AggregatedState> aggregatedStates =
+						bssidToAggregatedStates
+							.computeIfAbsent(key, k -> new ArrayList<>());
+					AggregatedState aggState =
+						new AggregatedState(association, radioInfo);
+
+					boolean toBeAggregated = false;
+					for (
+						AggregatedState oldAggregatedState : aggregatedStates
+					) {
+						if (oldAggregatedState.add(aggState)) {
+							toBeAggregated = true;
+							break;
+						}
+					}
+					if (!toBeAggregated) {
+						aggregatedStates.add(aggState);
+					}
+					bssidToAggregatedStates.put(key, aggregatedStates);
+				}
+			}
+		}
+	}
+
+	/**
+	 * This method aggregates States by bssid&station key pair and radio info.
+	 * if two States of the same bssid&station match in channel, channel width and tx_power
+	 * need to be aggregated to one {@code AggregatedState}. Currently only mcs and
+	 * rssi fields are being aggregated. They are of List<Integer> type in AggregatedState,
+	 * which list all the values over the time.
+	 *
+	 * @param dataModel the data model which includes the latest recorded States.
+	 * @param obsoletionPeriodMs the maximum amount of time (in milliseconds) it
+	 *                           is worth aggregating over, starting from the
+	 *                           most recent States and working backwards in time.
+	 * 							 A State exactly {@code obsoletionPeriodMs} ms earlier
+	 * 							 than the most recent State is considered non-obsolete
+	 *                           (i.e., the "non-obsolete" window is inclusive).
+	 *                           Must be non-negative.
+	 * @param refTimeMs	the reference time were passed to make testing easier
+	 * @return  Map<String, Map<String, List<AggregatedState>>> A map from serial number to
+	 * 							a map from bssid_station String pair to a list of AggregatedState.
+	 */
+	public static Map<String, Map<String, List<AggregatedState>>> getAggregatedStates(
+		Modeler.DataModel dataModel,
+		long obsoletionPeriodMs,
+		long refTimeMs
+	) {
+		if (obsoletionPeriodMs < 0) {
+			throw new IllegalArgumentException(
+				"obsoletionPeriodMs must be non-negative."
+			);
+		}
+		Map<String, Map<String, List<AggregatedState>>> aggregatedStates =
+			new HashMap<>();
+
+		for (
+			Map.Entry<String, List<State>> deviceToStateList : dataModel.latestStates
+				.entrySet()
+		) {
+			String serialNumber = deviceToStateList.getKey();
+			List<State> states = deviceToStateList.getValue();
+
+			if (states.isEmpty()) {
+				continue;
+			}
+
+			/**
+			 * Sort in reverse chronological order. Sorting is done just in case the
+			 * States in the original list are not chronological already - although
+			 * they are inserted chronologically, perhaps latency, synchronization, etc.
+			 */
+			states.sort(
+				(state1, state2) -> -Long.compare(state1.unit.localtime, state2.unit.localtime)
+			);
+
```

Comment:
So the first one rather than the last one should be latest, right? but you retrieve the last one as latest in all other places.

Reply:
The first one is the most recent. And she starts iterating from the beginning, so it's fine, right?

Reply:
But `getLatestState` in this file selected the last one as "latest".